File: src/main/ruby/g0001_0100/s0006_zigzag_conversion/solution.rb
# #Medium #String #Top_Interview_150_Array/String
# #2023_11_14_Time_77_ms_(100.00%)_Space_211_MB_(89.81%)

# @param {String} s
# @param {Integer} num_rows
# @return {String}
def convert(s, num_rows)
  s_len = s.length
  return s if num_rows == 1

  max_dist = num_rows * 2 - 2
  buf = ''

  (0...num_rows).each do |i|
    index = i

    if i.zero? || i == num_rows - 1
      while index < s_len
        buf << s[index]
        index += max_dist
      end
    else
      while index < s_len
        buf << s[index]
        index += max_dist - i * 2

        break if index >= s_len

        buf << s[index]
        index += i * 2
      end
    end
  end

  buf
end

File: src/main/dart/g0001_0100/s0006_zigzag_conversion/Solution.dart
// #Medium #String #Top_Interview_150_Array/String
// #2024_09_30_Time_374_ms_(94.79%)_Space_150.8_MB_(57.29%)

class Solution {
  String convert(String s, int numRows) {
    int sLen = s.length;
    if (numRows == 1) {
      return s;
    }
    int maxDist = numRows * 2 - 2;
    StringBuffer buf = StringBuffer();

    for (int i = 0; i < numRows; i++) {
      int index = i;
      if (i == 0 || i == numRows - 1) {
        while (index < sLen) {
          buf.write(s[index]);
          index += maxDist;
        }
      } else {
        while (index < sLen) {
          buf.write(s[index]);
          index += maxDist - i * 2;
          if (index >= sLen) {
            break;
          }
          buf.write(s[index]);
          index += i * 2;
        }
      }
    }
    return buf.toString();
  }
}
